fw/http_msg.c

```diff
@@ -1313,115 +1329,253 @@ tfw_http_msg_expand_data(TfwMsgIter *it, struct sk_buff **skb_head,
 }
 
 static int
-tfw_h2_msg_rewrite_data_common(TfwHttpTransIter *mit, const TfwStr *str,
-			       const char *stop,
-			       void cpy(void *dest, const void *src, size_t n))
+tfw_http_msg_alloc_from_pool(TfwHttpTransIter *mit, TfwPool* pool, size_t size)
 {
-	const TfwStr *c, *end;
+	int r;
+	bool np;
+	char* addr;
 	TfwMsgIter *it = &mit->iter;
-	TfwNextHdrOp *next = &mit->next;
+	struct skb_shared_info *si = skb_shinfo(it->skb);
 
-	BUG_ON(!stop);
-	BUG_ON(TFW_STR_DUP(str));
+	addr = tfw_pool_alloc_not_align_np(pool, size, &np);
+	if (!addr)
+		return -ENOMEM;
 
-	TFW_STR_FOR_EACH_CHUNK(c, str, end) {
-		const char *addr, *next_ptr;
-		long offset, stop_offset;
-		unsigned int c_off = 0, f_size, c_size, f_room, n_copy;
-this_chunk:
-		c_size = c->len - c_off;
-		if (it->frag >= 0) {
-			skb_frag_t *frag = &skb_shinfo(it->skb)->frags[it->frag];
+	if (np) {
+		r = ss_skb_add_frag(it->skb_head, it->skb, addr, ++it->frag,
+				    size);
+		if (unlikely(r))
+			return r;
+	} else {
+		skb_frag_size_add(&si->frags[it->frag], size);
+	}
 
-			f_size = skb_frag_size(frag);
-			addr = skb_frag_address(frag);
-		} else {
-			f_size = skb_headlen(it->skb);
-			addr = it->skb->data;
-		}
+	ss_skb_adjust_data_len(it->skb, size);
+	mit->curr_ptr = addr;
 
-		offset = mit->curr_ptr - addr;
-		if (WARN_ON_ONCE(offset < 0 || offset >= f_size))
-			return -EINVAL;
+	return 0;
+}
 
-		f_room = f_size - offset;
-		n_copy = min(c_size, f_room);
-		next_ptr = mit->curr_ptr + n_copy;
+/**
+ * Add first paged fragment using TfwPool and reserve room for HEADERS frame.
+ *
+ * This function must be used as start point of message transformation. After
+ * calling this you must use @pool only for allocating paged fragments during
+ * message trasformation. After it @pool can be used for anything else.
```

Comment:
I didn't get
> After calling this you must use @pool only for allocating paged fragments during message trasformation. After it @pool can be used for anything else.

Why? What happened if after the first paged fragment allocation we say reallocate some TfwStr? What "after it" means exactly - single paged fragment allocation?

Reply:
For fragments we trying to keep memory continuous. If we allocate some TfwStr after first paged fragment allocation we break continuous. 

For instance:  We have 2kb http headers, these headers have to be encoded into single paged fragment which is continuous and pool has enough space in current chunk. If we allocate TfwStr in the middle of encoding process, we got a gap between data, which will split the paged fragment. As a result we have 2 paged fragments.